src/ATen/native/xpu/sycl/MultinomialKernel.cpp

```diff
@@ -0,0 +1,533 @@
+         curDist += item.get_group_range(0)) {
+      // First pass, find the total sum of the distribution
+      accscalar_t sum = accZero;
+      scalar_t val;
+      for (int cat = item.get_local_id(0); cat < categories_;
+           cat += item.get_local_range(0)) {
+        val = dist_[curDist * stride_dist_ + cat * stride_categories_];
+        SYCL_KERNEL_ASSERT(!at::_isnan(val));
+        SYCL_KERNEL_ASSERT(!_isinf(val));
+        SYCL_KERNEL_ASSERT(!(val < zero));
+        sum = sum + static_cast<accscalar_t>(val);
+      }
+
+      sum = GroupReduceSumSGSizeEqualstoNumSG(item, sum, smem);
+
+      // Broadcast sum and sample value
+      if (item.get_local_id(0) == 0) {
+        // Make sure the sum of our distribution didn't overflow
+        SYCL_KERNEL_ASSERT(!_isinf(val));
+        SYCL_KERNEL_ASSERT(sum > accZero);
+
+        foundPos_[0] = 0;
+        smem[0] = sum;
+        smem[1] = sampled_[curDist];
+      }
+      item.barrier(sycl_local_fence);
+
+      sum = smem[0];
+      scalar_t sample = static_cast<scalar_t>(smem[1]);
+      item.barrier(sycl_local_fence);
+
+      if (sum == accZero) {
+        // Choose the first element
+        if (local_id == 0) {
+          dest_[curDist] = 0;
+        }
+
+        continue;
+      }
+
+      int chunks = (categories_ + (int)local_range - 1) / local_range;
+      accscalar_t prevHighProb = accZero;
+      found_[0] = false;
+
+      for (int chunk = 0; chunk < chunks && !found_[0]; ++chunk) {
+        // All threads in bounds load a value
+        int cat = chunk * local_range + local_id;
+
+        accscalar_t dist_val = cat < categories_
+            ? static_cast<accscalar_t>(
+                  dist_[curDist * stride_dist_ + cat * stride_categories_]) /
+                sum
+            : accZero;
+
+        smem[local_id] = dist_val;
+        item.barrier(sycl_local_fence);
+
+        // Perform an inclusive prefix sum of the shared memory contents
+        for (int offset = 1; offset < local_range; offset *= 2) {
+          accscalar_t val = accZero;
+
+          if (local_id >= offset) {
+            val = smem[local_id - offset] + smem[local_id];
+          }
+
+          item.barrier(sycl_local_fence);
+          if (local_id >= offset) {
+            smem[local_id] = val;
+          }
+          item.barrier(sycl_local_fence);
+        }
+
+        // Each thread will check to see if the sample falls in its
+        // bucket
+        scalar_t curBucket =
+            static_cast<scalar_t>(smem[local_id] + prevHighProb);
+        scalar_t prevBucket = static_cast<scalar_t>(
+            local_id == 0 ? prevHighProb : smem[local_id - 1] + prevHighProb);
+        bool inBucket = (cat < categories_) &&
+            (!(sample >= curBucket) && (sample >= prevBucket) &&
+             (dist_val > zero));
+
+        if (inBucket) {
+          // We're done; we have the sample
+          // Torch indices are 1-based
+          atomicMax(
```

Comment:
Ditto. SYCL supports atomic_ref::fetch_max for 32 bit data types. Please check.
<img width="604" alt="image" src="https://github.com/intel/torch-xpu-ops/assets/28250760/d4f73b63-1194-4cff-b0e7-f71069fffd0b">

Reply:
done